### ISSUE TITLE
Cancel redirection on login page

### ISSUE THREAD

**Issue by ayoung** [IdentityServer/IdentityServer3#2842]
- [x] I read and understood how to enable [logging](https://identityserver.github.io/Documentation/docsv2/configuration/logging.html)

### Question / Issue
I am required to provide a "cancel" button on the login view that redirects the user back to the originating client page. I am not able to do this because `LoginViewModel` does not have enough information for me to look up that information from the client store. I need `LoginViewModel` to have the `ClientId` in order to do that lookup.

Do you see this as an appropriate addition? Or is there possibly another way around this? For example, passing in a query param `cancel_redirect_uri` to the auth endpoint?

**Comment by brockallen** [comment]
You'd need to use the clieint's redirect_uri and pass an error param with "access_denied":

https://tools.ietf.org/html/rfc6749#section-4.1.2.1

**Comment by ayoung** [comment]
I'm having a hard time figuring out how/where to implement this logic. The cancel button would be on the login form. Clicking the cancel button would POST to `/login?signin=xxxxxxx`. If I try to implement this in my `UserService.AuthenticateLocalAsync`, there is no way to detect whether the current request is for a cancellation or for a normal login. the `LocalAuthenticationContext` does not provide that level of information for me to make that distinction. The context also does not provide the client's `redirect_uri`.

What should I be extending or modifying to get this behavior?

**Comment by brockallen** [comment]
> If I try to implement this in my UserService.AuthenticateLocalAsync(), there is no way to detect whether the current request is for a cancellation or for a normal login

You can inject the `OwinEnvironmentService` into your ctor to get custom form elements. The hard part is then doing a redirect... I think we might need to add an OWIN environment extension for this path (e.g. ReturnAuthorizationError or some such).

**Comment by slaneyrw** [comment]
Sorry for commenting on a closed issue, but has there been any progress on this.  I need to redirect back to the relying app if the user cancels the login.

I've tried returning an AuthenticationError result from a custom user service in preauth and implementing a custom request validator and returning error information but I cannot seem to get IdSvr to redirect back to the original app with the error and error_description.  It will only show the error screen in IDSvr.

The only process I have found that can redirect back to the app is a consent failure, but that requires an authenticated user.  Short of creating a "dummy" logged in user ( hack alert! ), is there any other way to get IDSvr to redirect back ?
